infrastructure/tari_util/src/ciphers/chacha20.rs

```diff
@@ -0,0 +1,416 @@
+{
+    fn seal(plain_text: &D, key: &Vec<u8>, nonce: &Vec<u8>) -> Result<Vec<u8>, CipherError> {
+        // Validation
+        if key.len() != 32 {
+            return Err(CipherError::KeyLengthError);
+        }
+        if nonce.len() != 12 {
+            return Err(CipherError::NonceLengthError);
+        }
+
+        let mut sized_key = [0; 32];
+        sized_key.copy_from_slice(key.as_slice());
+
+        let mut sized_nonce = [0; 12];
+        sized_nonce.copy_from_slice(nonce.as_slice());
+
+        let cipher_text = ChaCha20::encode_with_nonce(plain_text.as_bytes(), &sized_key, &sized_nonce);
+        // Clear copied private data
+        sized_key.clear();
+        sized_nonce.clear();
+
+        Ok(cipher_text)
+    }
+
+    fn open(cipher_text: &Vec<u8>, key: &Vec<u8>, nonce: &Vec<u8>) -> Result<D, CipherError> {
+        // Validation
+        if key.len() != 32 {
+            return Err(CipherError::KeyLengthError);
+        }
+        if nonce.len() != 12 {
+            return Err(CipherError::NonceLengthError);
+        }
+
+        let mut sized_key = [0; 32];
+        sized_key.copy_from_slice(key.as_slice());
+
+        let mut sized_nonce = [0; 12];
+        sized_nonce.copy_from_slice(nonce.as_slice());
+
+        let plain_text = ChaCha20::decode_with_nonce(cipher_text, &sized_key, &sized_nonce);
+        // Clear copied private data
+        sized_key.clear();
+        sized_nonce.clear();
+
+        Ok(D::from_vec(&plain_text)?)
+    }
+
+    fn seal_with_integral_nonce(plain_text: &D, key: &Vec<u8>) -> Result<Vec<u8>, CipherError> {
+        // Validation
+        if key.len() != 32 {
+            return Err(CipherError::KeyLengthError);
+        }
+
+        let mut sized_key = [0; 32];
+        sized_key.copy_from_slice(key.as_slice());
+
+        let mut rng = OsRng::new().unwrap();
+        let mut nonce = [0u8; 12];
+        rng.fill_bytes(&mut nonce);
+
+        let cipher_text = ChaCha20::encode_with_nonce(plain_text.as_bytes(), &sized_key, &nonce);
+        let mut nonce_with_cipher_text: Vec<u8> = nonce.to_vec();
+        nonce_with_cipher_text.extend(cipher_text);
+
+        // Clear copied private data
+        sized_key.clear();
+        nonce.clear();
+
+        Ok(nonce_with_cipher_text)
+    }
+
+    fn open_with_integral_nonce(cipher_text: &Vec<u8>, key: &Vec<u8>) -> Result<D, CipherError> {
+        // Validation
+        if key.len() != 32 {
+            return Err(CipherError::KeyLengthError);
+        }
+        // If the cipher text is shorter than the required nonce length then the nonce is not properly included
+        if cipher_text.len() < 12 {
+            return Err(CipherError::NonceLengthError);
+        }
+
+        let mut sized_key = [0; 32];
+        sized_key.copy_from_slice(key.as_slice());
+
+        let mut nonce = [0u8; 12];
+        nonce.copy_from_slice(&cipher_text.clone()[0..12]);
+
+        let plain_text = ChaCha20::decode_with_nonce(&cipher_text[12..], &sized_key, &nonce);
```

Comment:
I might be wrong, but won't `cipher_text[12..]` panic if the cipher_text only has a nonce and no text.  Then the cipher_text will have 12 bytes and only index 0 to 11 will be valid.

Reply:
true, enforced that there must be at least 1 byte of data during encryption and decryption